# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContext.cpp
# Commit: 38b03d99f39c
# Full Hash: 38b03d99f39cc445cf164cfcc9c6bfe07d26c3ec
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2019-05-25 09:39:13
# Description:
#   Bug 1535835 - Remove global weak ref to BrowsingContext within Unlink(), r=farre
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D30683
# ==============================================================================

diff -r 15293a40e833 -r 38b03d99f39c docshell/base/BrowsingContext.cpp
--- a/docshell/base/BrowsingContext.cpp	Fri May 24 22:36:49 2019 +0000
+++ b/docshell/base/BrowsingContext.cpp	Fri May 24 20:15:53 2019 +0000
@@ -50,16 +50,12 @@
 
 static LazyLogModule gBrowsingContextLog("BrowsingContext");
 
-template <template <typename> class PtrType>
-using BrowsingContextMap =
-    HashMap<uint64_t, PtrType<BrowsingContext>, DefaultHasher<uint64_t>,
-            InfallibleAllocPolicy>;
+typedef nsDataHashtable<nsUint64HashKey, BrowsingContext*> BrowsingContextMap;
 
-static StaticAutoPtr<BrowsingContextMap<WeakPtr>> sBrowsingContexts;
+static StaticAutoPtr<BrowsingContextMap> sBrowsingContexts;
 
 static void Register(BrowsingContext* aBrowsingContext) {
-  MOZ_ALWAYS_TRUE(
-      sBrowsingContexts->putNew(aBrowsingContext->Id(), aBrowsingContext));
+  sBrowsingContexts->Put(aBrowsingContext->Id(), aBrowsingContext);
 
   aBrowsingContext->Group()->Register(aBrowsingContext);
 }
@@ -75,7 +71,7 @@
 /* static */
 void BrowsingContext::Init() {
   if (!sBrowsingContexts) {
-    sBrowsingContexts = new BrowsingContextMap<WeakPtr>();
+    sBrowsingContexts = new BrowsingContextMap();
     ClearOnShutdown(&sBrowsingContexts);
   }
 }
@@ -85,11 +81,7 @@
 
 /* static */
 already_AddRefed<BrowsingContext> BrowsingContext::Get(uint64_t aId) {
-  if (BrowsingContextMap<WeakPtr>::Ptr abc = sBrowsingContexts->lookup(aId)) {
-    return do_AddRef(abc->value().get());
-  }
-
-  return nullptr;
+  return do_AddRef(sBrowsingContexts->Get(aId));
 }
 
 CanonicalBrowsingContext* BrowsingContext::Canonical() {
@@ -361,7 +353,7 @@
 bool BrowsingContext::IsCached() { return Group()->IsContextCached(this); }
 
 bool BrowsingContext::HasOpener() const {
-  return sBrowsingContexts->has(mOpenerId);
+  return sBrowsingContexts->Contains(mOpenerId);
 }
 
 void BrowsingContext::GetChildren(Children& aChildren) {
@@ -528,7 +520,7 @@
   MOZ_DIAGNOSTIC_ASSERT(!mGroup || !mGroup->IsContextCached(this));
 
   if (sBrowsingContexts) {
-    sBrowsingContexts->remove(Id());
+    sBrowsingContexts->Remove(Id());
   }
 }
 
@@ -569,6 +561,10 @@
 NS_IMPL_CYCLE_COLLECTION_CLASS(BrowsingContext)
 
 NS_IMPL_CYCLE_COLLECTION_UNLINK_BEGIN(BrowsingContext)
+  if (sBrowsingContexts) {
+    sBrowsingContexts->Remove(tmp->Id());
+  }
+
   NS_IMPL_CYCLE_COLLECTION_UNLINK(mDocShell, mChildren, mParent, mGroup,
                                   mEmbedderElement)
   if (XRE_IsParentProcess()) {